### DISCUSSION TITLE
Remotely starting a stored program on the PU Hub

### DISCUSSION THREAD

**Discussion by Novakasa** [pybricks/support#282] [Q&A] [answered]
Hey all, thanks for the amazing work on this project!
Right now I am using pybricksdev as part of a larger project that runs multiple Poweredup Hubs that have custom programs running on them. Right now, starting the program on the hub takes a long time (5-10 seconds) per hub, which I assume is due to having to store the program on the hub before running it. Since it is possible to store a program on a PoweredUp Hub when flashing the firmware, is there a way with pybricksdev to remotely start the existing program? The `BLEPUPConnection.run` method currently mandatorily needs a `py_path` argument. I would have tried to hack it myself, but looking at the code of `BLEPUPConnection.run` I don't easily see the code that tells the hub to just start the copied program.
Any insight on this would be appreciated!

**Comment by dlech** [comment, marked as answer]
Downloading and running a program should take less that a second unless the program is just really huge (1000+ lines).

There isn't currently a way to remotely start the program that gets flashed with the firmware, but that is a good feature request.

**Reply by Novakasa** [reply]
Oh, that's interesting, the program is only about 200 lines, so I will go looking for the reason why starting seems to be so slow for me.

**Reply by Novakasa** [reply]
Alright, so starting the program using this script:
```python
import time
import asyncio
from pybricksdev.connections import BLEPUPConnection
from pybricksdev.ble import find_device

async def main_coro():
    device = await find_device("Pybricks Hub")
    hub = BLEPUPConnection()
    await hub.connect(device)
    t0 = time.time()
    await hub.run("autonomous_train.py", wait=False)
    t = time.time()-t0
    print(f"starting program took {t} seconds!")

asyncio.run(main_coro())
```
I get as output:
```
Searching for Pybricks Hub
Connecting to 90:84:2B:BE:1E:40: Pybricks Hub
Connected successfully!
starting program took 9.359002113342285 seconds!
```
Which is a repeatable result with both of my City-hubs. The script I am running is 214 lines. I am not an expert on bluetooth, could this be so slow because I am using a very old bluetooth dongle (must be from around 2011)?

**Reply by dlech** [reply]
I don't think the dongle has anything to do with it. I have a feeling that there is probably something that could be improved in `pybricksdev`.

**Reply by Novakasa** [reply]
Alright, good to know. Thanks for the answers!
